text/1162449/undirected_ring_stages_with_feedback.tex

```diff
@@ -0,0 +1,99 @@
+(node that is not "\textit{passive}" will be also regarded as "\textit{active}")\\
+\textit{E-candidate} - node is taking part in this electoral round\\ \textit{A-candidate} - node awaits feedback from its active neighbours\\ \textit{Leader} - node is the leader\\
+\\
+Messages sent will consist of pairs of "type" and "value". Values will be from the set of unique identifiers of the nodes participating, while type can be  either \textit{E-msg}, \textit{A-msg} or \textit{T-msg}.\\
+By  \textit{left}, \textit{right} and  \textit{both}, we will denote directions in witch node sends or from which it receives a message.\\
+By $i$ we will denote the node's unique identifier and by $n(i)$ the node whose unique identifier is $i$.
+\\
+Upon initialization each node starts with status \textit{E-candidate}.\\
+\\
+Each round starts with each \textit{E-candidate} sending message (\textit{E-msg},$i$) in both of its directions.\\
+\\
+Then, each \textit{E-candidate} awaits to receive a message from both of its directions.\\
+Let's denote them as $(\textit{E-msg},vl)$ received from \textit{left} and $(\textit{E-msg},vr)$ received from \textit{right}.\\
+Let $y=max(i,vl,vr)$. \\
+If $y>i$, then $n(i)$ will send $(\textit{A-msg},y)$ to left if $y=vr$ or to right if $y=vl$.\\
+Then regardless if any message was sent from $n(i)$, it will change its state to \textit{A-candidate}.\\
+If $i=vl=vr$ then $n(i)$ is the only \textit{E-candidate} in the ring. It should become the \textit{Leader} and send \textit{T-msg} in one of its directions.\\
+\\
+Each \textit{A-candidate} awaits approval from both of its active neighbours, meaning, it waits for message $(\textit{A-msg},i)$ from both sides. Afterwards, it changes its status to \textit{E-candidate} to participate in the next round.\\
+If \textit{A-candidate} receives $(\textit{E-msg},i)$, it should treat it as a negative feedback, it should become \textit{passive} and pass the received message.\\
+\\
+\textit{passive} node passes every message, except for potential $(\textit{A-msg},i)$ which is no longer needed.\\
+If it receives $(\textit{T-msg},val)$, it learns that $n(val)$ is the leader, passes the message and finishes.\\
+\\
+\textit{Leader} finishes when it receives $(\textit{T-msg},i)$.\\
+\\
+It shall be noted that receiving \textit{T-msg} can be used to obtain common sense of direction in the ring.
+\newpage
+\subsection*{Correctness}
+Let $j=max\{i | n(i) \text{ participates in the algorithm}\}$. \\
+Then from the voting and checking procedure follows that at each electoral round $n(j)$ will defeat both of its immediate active neighbours; hence the number of \textit{passive} nodes increases monotonically. At the same time $n(j)$ will never be defeated (receive negative feedback). This implies that at some round $n(j)$ will be the only active node and both of its \textit{E-msg} messages will return to it, making it the leader.\\
```

Comment:
Please use line breaks sparingly, only at the last resort. The proper way to split text is to leave blank lines, which forces LaTeX to render separate subgraphs. Please remove all \\ from this file to practice it.